File: 02_Libraries/go/kirin_dns.go
// Package kirindns implements the KirinDNS Resolution Protocol (ADRP) v2.0.
//
// Architecture:
//
//	SRV records for service port discovery (_kirinnet-http._tcp, etc.)
//	TXT records for identity metadata (id=;key=;nick=;ipfs=)
//
// Example:
//
//	srv, err := kirindns.ResolveService("alice.kirinnet.org", "ws")
//	fmt.Printf("WS: %s:%d\n", srv.Target, srv.Port)
//
//	id, err := kirindns.ResolveIdentity("alice.kirinnet.org")
//	fmt.Printf("ID: %s, Nick: %s\n", id.ID, id.Nick)
package kirindns

import (
	"context"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"net"
	"strings"
	"time"
)

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

// SRVResult holds a resolved SRV service target and port.
type SRVResult struct {
	Target string
	Port   uint16
}

// Identity holds parsed identity metadata from a TXT record.
type Identity struct {
	ID   string
	Key  string
	Nick string // optional
	IPFS bool   // optional, defaults to false
}

// ---------------------------------------------------------------------------
// did:dns three-record identity model (spec §3.2.1 / did-dns-protocol §2)
// ---------------------------------------------------------------------------

// DidDnsIdentity holds a parsed did:dns identity (declaration + public key,
// optional blacklist). Trust policy is applied by the caller via IsValid
// (fail-closed default): version 1 + ed25519 + fingerprint chain + not revoked
// + not expired.
type DidDnsIdentity struct {
	Version       int
	Fingerprint   string
	Nickname      string // Base64URL(UTF-8), may be empty
	Gender        string // M/F/O/X, may be empty
	IssuedAt      int64  // Unix seconds, 0 if absent
	ExpiresAt     int64
	KeyType       string
	PublicKeyB64  string
	Blacklist     []string
	RawDeclaration string
	RawPublicKey   string
}

const (
	didDnsPrefix       = "did:dns:"
	didDnsDeclPrefix   = "did:dns:v="
	didDnsPkPrefix     = "did:dns:pk;"
	didDnsBlackPrefix  = "did:dns:black;"
	didDnsKtyEd25519   = "ed25519"
	didDnsFingerprintBytes = 12
	didDnsFreshnessWindow  = 5 * 60 // ±5 minutes (spec §3.2.1)
)

// PublicKeyBytes decodes the Base64URL public key.
func (d *DidDnsIdentity) PublicKeyBytes() ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(d.PublicKeyB64)
}

// ComputeFingerprint recomputes fp = Base64URL(SHA-256(pk)[0:12]).
func (d *DidDnsIdentity) ComputeFingerprint() (string, error) {
	pk, err := d.PublicKeyBytes()
	if err != nil {
		return "", err
	}
	digest := sha256.Sum256(pk)
	return base64.RawURLEncoding.EncodeToString(digest[:didDnsFingerprintBytes]), nil
}

// FingerprintChainOk reports whether the declared fp matches the recomputed fp.
func (d *DidDnsIdentity) FingerprintChainOk() bool {
	if d.Fingerprint == "" {
		return false
	}
	got, err := d.ComputeFingerprint()
	return err == nil && got == d.Fingerprint
}

// IsRevoked reports whether the declaration fingerprint appears in the blacklist.
func (d *DidDnsIdentity) IsRevoked() bool {
	for _, f := range d.Blacklist {
		if f == d.Fingerprint {
			return true
		}
	}
	return false
}

// IsExpired reports whether exp <= now (0 ExpiresAt means no expiry check).
func (d *DidDnsIdentity) IsExpired(now int64) bool {
	if d.ExpiresAt == 0 {
		return false
	}
	return now >= d.ExpiresAt
}

// IsStale reports whether iat is outside ±5min of now.
func (d *DidDnsIdentity) IsStale(now int64) bool {
	if d.IssuedAt == 0 {
		return false
	}
	diff := now - d.IssuedAt
	if diff < 0 {
		diff = -diff
	}
	return diff > didDnsFreshnessWindow
}

// IsValid is the composite policy check (fail-closed).
func (d *DidDnsIdentity) IsValid(now int64) bool {
	return d.Version == 1 &&
		d.KeyType == didDnsKtyEd25519 &&
		d.FingerprintChainOk() &&
		!d.IsRevoked() &&
		!d.IsExpired(now)
}

// NicknameDecoded decodes the Base64URL(UTF-8) nickname, or "" if absent/invalid.
func (d *DidDnsIdentity) NicknameDecoded() string {
	if d.Nickname == "" {
		return ""
	}
	b, err := base64.RawURLEncoding.DecodeString(d.Nickname)
	if err != nil {
		return ""
	}
	return string(b)
}

// ParseDidDnsIdentity classifies TXT records by did:dns: sub-type and assembles
// an identity. Returns nil if no did:dns records or declaration+pk are missing.
// Pure function (no network) — usable for tests.
func ParseDidDnsIdentity(txtRecords []string) *DidDnsIdentity {
	var declRaw, pkRaw, blackRaw string
	for _, raw := range txtRecords {
		s := strings.TrimSpace(raw)
		switch {
		case strings.HasPrefix(s, didDnsDeclPrefix) && declRaw == "":
			declRaw = s
		case strings.HasPrefix(s, didDnsPkPrefix) && pkRaw == "":
			pkRaw = s
		case strings.HasPrefix(s, didDnsBlackPrefix) && blackRaw == "":
			blackRaw = s
		}
	}
	if declRaw == "" || pkRaw == "" {
		return nil
	}

	decl := parseDidDnsKv(strings.TrimPrefix(declRaw, didDnsPrefix))
	pk := parseDidDnsKv(strings.TrimPrefix(pkRaw, didDnsPrefix))

	id := &DidDnsIdentity{
		Version:        didDnsAtoi(decl["v"], 1),
		Fingerprint:    decl["fp"],
		Nickname:       decl["n"],
		Gender:         decl["g"],
		KeyType:        pk["kty"],
		PublicKeyB64:   pk["pk"],
		RawDeclaration: declRaw,
		RawPublicKey:   pkRaw,
	}
	if pk["kty"] == "" {
		id.KeyType = didDnsKtyEd25519
	}
	if v, ok := decl["iat"]; ok {
		id.IssuedAt = didDnsAtoi64(v, 0)
	}
	if v, ok := decl["exp"]; ok {
		id.ExpiresAt = didDnsAtoi64(v, 0)
	}

	if blackRaw != "" {
		black := parseDidDnsKv(strings.TrimPrefix(blackRaw, didDnsPrefix))
		for _, f := range strings.Split(black["fp"], ",") {
			if f != "" {
				id.Blacklist = append(id.Blacklist, f)
			}
		}
	}
	return id
}

func parseDidDnsKv(text string) map[string]string {
	out := map[string]string{}
	for _, pair := range strings.Split(text, ";") {
		eq := strings.Index(pair, "=")
		if eq < 0 {
			continue
		}
		out[strings.TrimSpace(pair[:eq])] = strings.TrimSpace(pair[eq+1:])
	}
	return out
}

func didDnsAtoi(s string, def int) int {
	n := 0
	for _, c := range s {
		if c < '0' || c > '9' {
			return def
		}
		n = n*10 + int(c-'0')
	}
	if s == "" {
		return def
	}
	return n
}

func didDnsAtoi64(s string, def int64) int64 {
	var n int64
	seen := false
	for _, c := range s {
		if c < '0' || c > '9' {
			return def
		}
		n = n*10 + int64(c-'0')
		seen = true
	}
	if !seen {
		return def
	}
	return n
}

// FullResolution is the legacy wrapper result containing all services + identity.
type FullResolution struct {
	Domain   string
	WS       *SRVResult
	HTTP     *SRVResult
	HTTPS    *SRVResult
	Identity *Identity
}

// ResolvedPorts holds service port mappings (v1 legacy type).
type ResolvedPorts struct {
	HTTP, HTTPS, WS, WSS uint16
}

// ---------------------------------------------------------------------------
// Constants (spec Section 2.2)
// ---------------------------------------------------------------------------

// SRV service name prefixes.
const (
	SRVServiceHTTP  = "kirinnet-http"
	SRVServiceHTTPS = "kirinnet-https"
	SRVServiceWS    = "kirinnet-ws"
	SRVProto        = "tcp"
)

// Recognized service keys for resolveService.
var recognizedServices = map[string]string{
	"http":  SRVServiceHTTP,
	"https": SRVServiceHTTPS,
	"ws":    SRVServiceWS,
}

// Fallback ports (spec Section 3.3.1, Step 4)
const (
	DefaultHTTP  = 80
	DefaultHTTPS = 443
	DefaultWS    = 80
	DefaultWSS   = 443
)

// DefaultTimeout is the DNS query timeout.
var DefaultTimeout = 5 * time.Second

// ---------------------------------------------------------------------------
// Legacy v1 API (kept for backward compatibility)
// ---------------------------------------------------------------------------

// Fallback returns a ResolvedPorts with all standard IANA fallback values.
func Fallback() ResolvedPorts {
	return ResolvedPorts{DefaultHTTP, DefaultHTTPS, DefaultWS, DefaultWSS}
}

// Resolve queries DNS and returns resolved ports (v1 TXT-based API).
// Deprecated: Use ResolveService and ResolveIdentity for v2 SRV-based resolution.
func Resolve(domain string) (ResolvedPorts, error) {
	return ResolveWithResolver(domain, "")
}

// ResolveWithResolver uses a custom DNS resolver address (v1 API).
// Deprecated: Use ResolveService and ResolveIdentity for v2 SRV-based resolution.
func ResolveWithResolver(domain, resolverAddr string) (ResolvedPorts, error) {
	ctx, cancel := context.WithTimeout(context.Background(), DefaultTimeout)
	defer cancel()
	return resolveV1(ctx, domain, resolverAddr)
}

func resolveV1(ctx context.Context, domain, resolverAddr string) (ResolvedPorts, error) {
	ports := Fallback()
	if !strings.HasSuffix(domain, ".") {
		domain += "."
	}

	var r *net.Resolver
	if resolverAddr != "" {
		r = &net.Resolver{
			PreferGo: true,
			Dial: func(ctx context.Context, network, addr string) (net.Conn, error) {
				d := net.Dialer{Timeout: DefaultTimeout}
				return d.DialContext(ctx, network, resolverAddr)
			},
		}
	} else {
		r = net.DefaultResolver
	}

	txts, err := r.LookupTXT(ctx, domain)
	if err != nil {
		return ports, nil
	}

	for _, txt := range txts {
		parsed, ok := parseTxtV1(txt)
		if !ok {
			continue
		}
		if v, e := parsed["http"]; e {
			ports.HTTP = v
		}
		if v, e := parsed["https"]; e {
			ports.HTTPS = v
		}
		if v, e := parsed["ws"]; e {
			ports.WS = v
		}
		if v, e := parsed["wss"]; e {
			ports.WSS = v
		}
		return ports, nil
	}
	return ports, nil
}

// ---------------------------------------------------------------------------
// Service Resolution (SRV) — v2 API
// ---------------------------------------------------------------------------

// ResolveService resolves a single service port via SRV.
//
// service must be one of: "http", "https", "ws".
// Returns nil, nil if no SRV record is found.
func ResolveService(domain, service string) (*SRVResult, error) {
	srvPrefix, ok := recognizedServices[service]
	if !ok {
		return nil, &ServiceError{Service: service}
	}

	ctx, cancel := context.WithTimeout(context.Background(), DefaultTimeout)
	defer cancel()

	_, addrs, err := net.DefaultResolver.LookupSRV(ctx, srvPrefix, SRVProto, domain)
	if err != nil {
		return nil, nil // no SRV record
	}
	if len(addrs) == 0 {
		return nil, nil
	}

	return selectBestSrv(addrs), nil
}

// srvLike is the subset of net.SRV used by the selection algorithm; tests can
// construct these directly to cover RFC 2782 priority/weight selection without
// hitting the network.
type srvLike struct {
	Target   string
	Port     uint16
	Priority uint16
	Weight   uint16
}

// selectBestSrv applies RFC 2782 selection (lowest priority, then highest
// weight) and returns the chosen record as an *SRVResult. Exposed for tests.
func selectBestSrv(addrs []*net.SRV) *SRVResult {
	if len(addrs) == 0 {
		return nil
	}
	best := addrs[0]
	for _, a := range addrs[1:] {
		if a.Priority < best.Priority {
			best = a
		} else if a.Priority == best.Priority && a.Weight > best.Weight {
			best = a
		}
	}
	target := strings.TrimSuffix(best.Target, ".")
	return &SRVResult{Target: target, Port: best.Port}
}

// ResolveAllServices resolves all SRV services for a domain.
//
// Returns a map of service name -> SRVResult (nil if not found).
func ResolveAllServices(domain string) map[string]*SRVResult {
	result := make(map[string]*SRVResult, len(recognizedServices))
	for svc := range recognizedServices {
		srv, err := ResolveService(domain, svc)
		if err != nil {
			srv = nil
		}
		result[svc] = srv
	}
	return result
}

// ---------------------------------------------------------------------------
// Identity Resolution (TXT) — v2 API
// ---------------------------------------------------------------------------

// parseIdentityTxt parses a semicolon-separated key=value TXT string.
//
// Format: id=<uuid>;key=<hex>;nick=<name>;ipfs=<bool>
// Returns nil if the string is not a valid identity record.
func parseIdentityTxt(txt string) *Identity {
	txt = strings.TrimSpace(txt)
	if txt == "" || !strings.HasPrefix(txt, "id=") {
		return nil
	}

	result := &Identity{}
	for _, pair := range strings.Split(txt, ";") {
		eq := strings.Index(pair, "=")
		if eq == -1 {
			continue
		}
		key := strings.TrimSpace(pair[:eq])
		val := strings.TrimSpace(pair[eq+1:])

		switch key {
		case "id":
			result.ID = val
		case "key":
			result.Key = val
		case "nick":
			result.Nick = val
		case "ipfs":
			result.IPFS = val == "true"
		}
	}

	if result.ID == "" || result.Key == "" {
		return nil
	}
	return result
}

// ResolveIdentity resolves identity metadata from TXT records.
//
// Returns nil, nil if no identity TXT record is found.
func ResolveIdentity(domain string) (*Identity, error) {
	ctx, cancel := context.WithTimeout(context.Background(), DefaultTimeout)
	defer cancel()

	txts, err := net.DefaultResolver.LookupTXT(ctx, domain)
	if err != nil {
		return nil, nil
	}

	for _, txt := range txts {
		id := parseIdentityTxt(txt)
		if id != nil {
			return id, nil
		}
	}
	return nil, nil
}

// ResolveIdentityDidDns resolves and assembles a did:dns identity for domain.
// Returns nil, nil on NXDOMAIN / no TXT / no did:dns records (fail-closed).
func ResolveIdentityDidDns(domain string) (*DidDnsIdentity, error) {
	ctx, cancel := context.WithTimeout(context.Background(), DefaultTimeout)
	defer cancel()

	txts, err := net.DefaultResolver.LookupTXT(ctx, domain)
	if err != nil {
		return nil, nil
	}
	return ParseDidDnsIdentity(txts), nil
}

// ---------------------------------------------------------------------------
// Legacy Compatibility Wrapper — v2
// ---------------------------------------------------------------------------

// ResolveKirinDNS performs full resolution: SRV + TXT + identity.
//
// This is the legacy wrapper. New code should use ResolveService and
// ResolveIdentity directly.
func ResolveKirinDNS(domain string) (*FullResolution, error) {
	wsSrv, _ := ResolveService(domain, "ws")
	if wsSrv == nil {
		wsSrv = &SRVResult{Target: domain, Port: DefaultWS}
	}

	httpSrv, _ := ResolveService(domain, "http")
	httpsSrv, _ := ResolveService(domain, "https")
	identity, _ := ResolveIdentity(domain)

	return &FullResolution{
		Domain:   domain,
		WS:       wsSrv,
		HTTP:     httpSrv,
		HTTPS:    httpsSrv,
		Identity: identity,
	}, nil
}

// ---------------------------------------------------------------------------
// Error types
// ---------------------------------------------------------------------------

// ServiceError is returned when an unrecognized service name is used.
type ServiceError struct {
	Service string
}

func (e *ServiceError) Error() string {
	return "unknown service: " + e.Service + ". Recognized: http, https, ws"
}

// ---------------------------------------------------------------------------
// v1 TXT parser (legacy, kept for backward compatibility)
// ---------------------------------------------------------------------------

var recognizedKeysV1 = map[string]bool{
	"http": true, "https": true, "ws": true, "wss": true,
}

func parseTxtV1(txt string) (map[string]uint16, bool) {
	txt = strings.TrimSpace(txt)

	var raw map[string]any
	if err := json.Unmarshal([]byte(txt), &raw); err != nil {
		return nil, false
	}

	result := make(map[string]uint16)
	hasKey := false
	for key, val := range raw {
		if !recognizedKeysV1[key] {
			continue
		}
		num, ok := val.(float64)
		if !ok || num < 1 || num > 65535 || num != float64(uint16(num)) {
			return nil, false
		}
		result[key] = uint16(num)
		hasKey = true
	}
	if !hasKey {
		return nil, false
	}
	return result, true
}

File: 02_Libraries/kotlin/KirinDns.kt
/**
 * KirinDNS Resolution Protocol (ADRP) v2.0 — Kotlin Client
 *
 * Implements ADRP v2.0 per 01_Standard/spec_v1.md and the did:dns three-record
 * identity model in 01_Standard/did-dns-protocol.md §2 (C-1 baseline, 2026-08-08).
 *
 * Architecture:
 *   SRV records (_kirinnet-http/https/ws._tcp) for service port discovery.
 *   TXT records for identity metadata in did:dns three-record form:
 *     did:dns:v=1;fp=<fp>;n=<nick>;g=<gender>;iat=<ts>;exp=<ts>   (declaration)
 *     did:dns:pk;kty=ed25519;pk=<pubkey-base64url>                 (public key)
 *     did:dns:black;fp=<fp1>,<fp2>,...                             (blacklist, optional)
 *   The tamper-evident fingerprint chain binds declaration to public key:
 *     fp == Base64URL(SHA-256(pk_bytes)[0:12])
 *
 * The v1 TXT-JSON port model (KirinPorts / parseTxt) is retained for backward
 * compatibility; new code MUST use resolveService / parseDidDnsIdentity.
 *
 * Pure Kotlin/JVM — uses javax.naming (JNDI DNS), no extra dependencies.
 *
 * Compile + run self-test:
 *   kotlinc KirinDns.kt -include-runtime -d KirinDns.jar
 *   kotlin KirinDns.jar
 */

package kirinnet

import java.security.MessageDigest
import java.util.Base64
import java.util.Hashtable
import javax.naming.directory.InitialDirContext

// ---------------------------------------------------------------------------
// Constants (spec §2.2, §3.3.1 Step 4, §3.2.1 / did-dns-protocol §2)
// ---------------------------------------------------------------------------

private val SRV_SERVICES = linkedMapOf(
    "http" to "_kirinnet-http._tcp",
    "https" to "_kirinnet-https._tcp",
    "ws" to "_kirinnet-ws._tcp",
)

const val DEFAULT_HTTP = 80
const val DEFAULT_HTTPS = 443
const val DEFAULT_WS = 80
const val DEFAULT_WSS = 443

private const val DID_DNS_PREFIX = "did:dns:"
private const val DID_DNS_DECL_PREFIX = "did:dns:v="
private const val DID_DNS_PK_PREFIX = "did:dns:pk;"
private const val DID_DNS_BLACK_PREFIX = "did:dns:black;"
private const val KTY_ED25519 = "ed25519"
private const val FRESHNESS_WINDOW_SECONDS = 5L * 60
private const val FINGERPRINT_BYTES = 12

// ---------------------------------------------------------------------------
// v2 result types
// ---------------------------------------------------------------------------

/** A resolved SRV service target. */
data class SrvResult(val target: String, val port: Int)

private data class SrvRecord(
    val target: String, val port: Int, val priority: Int, val weight: Int
)

/**
 * A parsed did:dns identity (declaration + public key, optional blacklist).
 * The caller applies trust policy via [isValid] (fail-closed).
 */
class DidDnsIdentity(
    var version: Int = 1,
    var fingerprint: String = "",
    var nickname: String? = null,        // Base64URL(UTF-8)
    var gender: String? = null,          // M/F/O/X
    var issuedAt: Long? = null,
    var expiresAt: Long? = null,
    var keyType: String = KTY_ED25519,
    var publicKeyB64Url: String = "",
    var blacklist: MutableList<String> = mutableListOf(),
    var rawDeclaration: String = "",
    var rawPublicKey: String = "",
) {
    val publicKeyBytes: ByteArray
        get() = base64UrlDecode(publicKeyB64Url)

    /** Recompute fp = Base64URL(SHA-256(pk)[0:12]). */
    fun computeFingerprint(): String {
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(publicKeyBytes)
        val truncated = digest.copyOfRange(0, FINGERPRINT_BYTES)
        return base64UrlEncode(truncated)
    }

    fun fingerprintChainOk(): Boolean =
        fingerprint.isNotEmpty() && fingerprint == computeFingerprint()

    fun isRevoked(): Boolean = blacklist.contains(fingerprint)

    fun isExpired(now: Long? = null): Boolean {
        val exp = expiresAt ?: return false
        val t = now ?: System.currentTimeMillis() / 1000L
        return t >= exp
    }

    fun isStale(now: Long? = null): Boolean {
        val iat = issuedAt ?: return false
        val t = now ?: System.currentTimeMillis() / 1000L
        return Math.abs(t - iat) > FRESHNESS_WINDOW_SECONDS
    }

    fun isValid(now: Long? = null): Boolean =
        version == 1
            && keyType == KTY_ED25519
            && fingerprintChainOk()
            && !isRevoked()
            && !isExpired(now)

    fun nicknameDecoded(): String? {
        val n = nickname ?: return null
        return try { String(base64UrlDecode(n), Charsets.UTF_8) } catch (_: Exception) { null }
    }

    override fun toString(): String =
        "DidDnsIdentity(fp=$fingerprint, kty=$keyType, chainOk=${fingerprintChainOk()})"
}

// ---------------------------------------------------------------------------
// v2: SRV service resolution
// ---------------------------------------------------------------------------

/**
 * Resolve a single service port via SRV.
 * @throws IllegalArgumentException on unknown service.
 */
fun resolveService(domain: String, service: String): SrvResult? {
    val srvPrefix = SRV_SERVICES[service]
        ?: throw IllegalArgumentException("Unknown service: $service. Recognized: http, https, ws")

    return try {
        val ctx = newInitialDirContext()
        val attrs = ctx.getAttributes("$srvPrefix.$domain", arrayOf("SRV"))
        val srvAttr = attrs.get("SRV") ?: return null

        val records = (0 until srvAttr.size()).mapNotNull {
            parseSrvRecord(srvAttr.get(it)?.toString())
        }
        if (records.isEmpty()) return null

        // RFC 2782: lowest priority, then highest weight
        val best = records.sortedWith(
            compareBy<SrvRecord> { it.priority }.thenByDescending { it.weight }
        ).first()
        val target = best.target.trimEnd('.')
        SrvResult(target, best.port)
    } catch (_: Exception) {
        null // NXDOMAIN / no SRV → null (fail-closed)
    }
}

private fun parseSrvRecord(s: String?): SrvRecord? {
    if (s == null) return null
    // JNDI SRV format: "priority weight port target"
    val parts = s.trim().split(Regex("\\s+"))
    if (parts.size < 4) return null
    return try {
        SrvRecord(parts[3], parts[2].toInt(), parts[0].toInt(), parts[1].toInt())
    } catch (_: NumberFormatException) {
        null
    }
}

// ---------------------------------------------------------------------------
// v2: did:dns identity resolution
// ---------------------------------------------------------------------------

/** Resolve and assemble a did:dns identity (fail-closed: null if absent). */
fun resolveIdentityDidDns(domain: String): DidDnsIdentity? {
    return try {
        val ctx = newInitialDirContext()
        val attrs = ctx.getAttributes(domain, arrayOf("TXT"))
        val txtAttr = attrs.get("TXT") ?: return null
        val records = (0 until txtAttr.size()).mapNotNull { i ->
            txtAttr.get(i)?.toString()?.let(::quoteStrip)
        }
        parseDidDnsIdentity(records)
    } catch (_: Exception) {
        null
    }
}

/**
 * Classify TXT records by did:dns: sub-type and assemble an identity.
 * Pure function (no network) — usable for tests.
 */
fun parseDidDnsIdentity(txtRecords: List<String>?): DidDnsIdentity? {
    if (txtRecords == null) return null
    var declRaw: String? = null
    var pkRaw: String? = null
    var blackRaw: String? = null

    for (raw in txtRecords) {
        val s = raw.trim()
        when {
            s.startsWith(DID_DNS_DECL_PREFIX) -> { if (declRaw == null) declRaw = s }
            s.startsWith(DID_DNS_PK_PREFIX) -> { if (pkRaw == null) pkRaw = s }
            s.startsWith(DID_DNS_BLACK_PREFIX) -> { if (blackRaw == null) blackRaw = s }
        }
    }
    if (declRaw == null || pkRaw == null) return null

    val decl = parseKv(declRaw.substring(DID_DNS_PREFIX.length))
    val pk = parseKv(pkRaw.substring(DID_DNS_PREFIX.length))

    val id = DidDnsIdentity(
        version = decl["v"]?.toIntOrNull() ?: 1,
        fingerprint = decl["fp"] ?: "",
        nickname = decl["n"]?.takeIf { it.isNotEmpty() },
        gender = decl["g"]?.takeIf { it.isNotEmpty() },
        issuedAt = decl["iat"]?.toLongOrNull(),
        expiresAt = decl["exp"]?.toLongOrNull(),
        keyType = pk["kty"] ?: KTY_ED25519,
        publicKeyB64Url = pk["pk"] ?: "",
        rawDeclaration = declRaw,
        rawPublicKey = pkRaw,
    )
    if (blackRaw != null) {
        val fpField = parseKv(blackRaw.substring(DID_DNS_PREFIX.length))["fp"] ?: ""
        id.blacklist = fpField.split(",").filter { it.isNotEmpty() }.toMutableList()
    }
    return id
}

// ---------------------------------------------------------------------------
// v1 legacy (deprecated)
// ---------------------------------------------------------------------------

/** @deprecated v1 TXT-JSON model. */
@Deprecated("v1 TXT-JSON model. Use resolveService + resolveIdentityDidDns.")
data class KirinPorts(
    val http: Int = 80,
    val https: Int = 443,
    val ws: Int = 80,
    val wss: Int = 443,
) {
    companion object { val FALLBACK = KirinPorts() }
}

/** @deprecated v1 TXT-JSON model. */
@Deprecated("v1 TXT-JSON model.")
object KirinDns {
    fun resolve(domain: String): KirinPorts {
        var ports = KirinPorts.FALLBACK
        return try {
            val ctx = newInitialDirContext()
            val attrs = ctx.getAttributes(domain, arrayOf("TXT"))
            val txtAttr = attrs.get("TXT") ?: return ports
            for (i in 0 until txtAttr.size()) {
                val parsed = parseTxt(txtAttr.get(i)?.toString())
                if (parsed != null) { ports = parsed; break }
            }
            ports
        } catch (_: Exception) {
            ports
        }
    }
}

/** @deprecated v1 TXT-JSON model. */
@Deprecated("v1 TXT-JSON model.")
fun parseTxt(txt: String?): KirinPorts? {
    if (txt.isNullOrBlank()) return null
    val trimmed = txt.trim()
    if (!trimmed.startsWith("{")) return null
    var http = 0; var https = 0; var ws = 0; var wss = 0; var found = 0
    for (key in listOf("http", "https", "ws", "wss")) {
        val search = "\"$key\":"
        var idx = trimmed.indexOf(search)
        if (idx < 0) continue
        idx += search.length
        while (idx < trimmed.length && (trimmed[idx] == ' ' || trimmed[idx] == '\t')) idx++
        var end = idx
        while (end < trimmed.length && trimmed[end].isDigit()) end++
        if (end == idx) continue
        val num = trimmed.substring(idx, end).toIntOrNull() ?: continue
        if (num < 1 || num > 65535) return null
        when (key) {
            "http" -> http = num
            "https" -> https = num
            "ws" -> ws = num
            "wss" -> wss = num
        }
        found++
    }
    if (found == 0) return null
    return KirinPorts(
        http = if (http > 0) http else DEFAULT_HTTP,
        https = if (https > 0) https else DEFAULT_HTTPS,
        ws = if (ws > 0) ws else DEFAULT_WS,
        wss = if (wss > 0) wss else DEFAULT_WSS,
    )
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

private fun newInitialDirContext(): InitialDirContext {
    val env = Hashtable<String, String>().apply {
        put("java.naming.factory.initial", "com.sun.jndi.dns.DnsContextFactory")
        put("java.naming.provider.url", "dns://")
    }
    return InitialDirContext(env)
}

private fun parseKv(text: String): Map<String, String> {
    val d = linkedMapOf<String, String>()
    for (pair in text.split(";")) {
        val eq = pair.indexOf('=')
        if (eq < 0) continue
        d[pair.substring(0, eq).trim()] = pair.substring(eq + 1).trim()
    }
    return d
}

/** JNDI sometimes wraps TXT values in double quotes; strip them. */
private fun quoteStrip(s: String): String {
    val t = s.trim()
    return if (t.length >= 2 && t.startsWith("\"") && t.endsWith("\"")) t.substring(1, t.length - 1) else t
}

// ---- Base64URL helpers (no padding, RFC 4648 §5) ----

private fun base64UrlEncode(bytes: ByteArray): String =
    Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

private fun base64UrlDecode(s: String): ByteArray {
    val pad = (4 - s.length % 4) % 4
    val padded = if (pad > 0) s + "====".substring(0, pad) else s
    return Base64.getUrlDecoder().decode(padded)
}

// ---------------------------------------------------------------------------
// Self-test (run: kotlin -cp KirinDns.jar kirinnet.MainKt  or  kotlin KirinDns.jar)
// ---------------------------------------------------------------------------
fun main() {
    // Deterministic 32-byte key + its fingerprint (matches JS/Python golden vector).
    val pkBytes = ByteArray(32) { it.toByte() }
    val pkB64 = base64UrlEncode(pkBytes)
    val sha = MessageDigest.getInstance("SHA-256")
    val fp = base64UrlEncode(sha.digest(pkBytes).copyOfRange(0, FINGERPRINT_BYTES))
    val now = 1_700_000_000L

    val recs = listOf(
        "v=spf1 include:_spf.kirinnet.org -all",
        "did:dns:v=1;fp=$fp;n=QWxpY2U;g=F;iat=$now;exp=${now + 3600}",
        "did:dns:pk;kty=ed25519;pk=$pkB64",
        "did:dns:black;fp=RevokedAaaa,RevokedBbbb",
    )

    val id = parseDidDnsIdentity(recs)
    check(id != null) { "did:dns identity must parse" }
    check(id!!.version == 1) { "version" }
    check(id.fingerprint == fp) { "fp" }
    check(id.keyType == "ed25519") { "kty" }
    check(id.fingerprintChainOk()) { "fingerprint chain" }
    check(id.isValid(now)) { "fresh ed25519 valid" }
    check(id.nicknameDecoded() == "Alice") { "nickname decode" }
    check(id.gender == "F") { "gender" }
    check(id.blacklist.size == 2) { "blacklist size" }
    check(!id.isRevoked()) { "not revoked" }

    // Tampered pk -> chain breaks
    val wrongPk = ByteArray(32) { 0xff.toByte() }
    val tampered = recs.toMutableList().apply { this[2] = "did:dns:pk;kty=ed25519;pk=${base64UrlEncode(wrongPk)}" }
    val broken = parseDidDnsIdentity(tampered)
    check(broken != null && !broken!!.fingerprintChainOk()) { "tampered pk breaks chain" }

    // Missing pk -> null
    check(parseDidDnsIdentity(listOf(recs[1])) == null) { "missing pk -> null" }
    // No did:dns -> null
    check(parseDidDnsIdentity(listOf("v=spf1 -all", "id=foo;key=bar")) == null) { "no did:dns -> null" }
    // RSA kty -> invalid
    val rsa = parseDidDnsIdentity(listOf(recs[1], "did:dns:pk;kty=rsa;pk=$pkB64"))
    check(rsa!!.keyType == "rsa" && !rsa.isValid(now)) { "rsa kty rejected" }

    // v1 legacy parser still works
    val p = parseTxt("""{"http":8080,"https":8443}""")
    check(p != null && p!!.http == 8080 && p.https == 8443) { "v1 parse" }
    check(parseTxt("{}") == null) { "v1 empty" }
    check(parseTxt("not json") == null) { "v1 not json" }

    // SRV resolution over network (best-effort; nonexistent -> null, no throw)
    check(resolveService("nonexistent.invalid", "ws") == null) { "nonexistent SRV null" }
    // Unknown service throws
    var threw = false
    try { resolveService("example.com", "wss") } catch (_: IllegalArgumentException) { threw = true }
    check(threw) { "unknown service should throw" }

    println("KirinDns Kotlin self-test: PASSED (incl. did:dns fingerprint chain)")
}
